In [15]:
import pandas as pd

df = pd.read_csv('raw_data.csv')
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)


In [ ]:
df.columns #(inventario)lista con los nombres de las columnas.

Index(['Unnamed: 0', 'age', 'attrition', 'businesstravel', 'dailyrate',
       'department', 'distancefromhome', 'education', 'educationfield',
       'employeecount', 'employeenumber', 'environmentsatisfaction', 'gender',
       'hourlyrate', 'jobinvolvement', 'joblevel', 'jobrole',
       'jobsatisfaction', 'maritalstatus', 'monthlyincome', 'monthlyrate',
       'numcompaniesworked', 'over18', 'overtime', 'percentsalaryhike',
       'performancerating', 'relationshipsatisfaction', 'standardhours',
       'stockoptionlevel', 'totalworkingyears', 'trainingtimeslastyear',
       'worklifebalance', 'yearsatcompany', 'yearsincurrentrole',
       'yearssincelastpromotion', 'yearswithcurrmanager',
       'sameasmonthlyincome', 'datebirth', 'salary', 'roledepartament',
       'numberchildren', 'remotework'],
      dtype='object')

In [ ]:
df.head()   #imprimir las primeras filas para hacerme una idea.



Unnamed: 0 age attrition businesstravel    dailyrate  \
0           0  51        No            NaN  2015.722222   
1           1  52        No            NaN  2063.388889   
2           2  42        No  travel_rarely  1984.253968   
3           3  47        No  travel_rarely  1771.404762   
4           4  46        No            NaN  1582.771346   

                 department  distancefromhome  education    educationfield  \
0                       NaN                 6          3               NaN   
1                       NaN                 1          4     Life Sciences   
2   Research & Development                  4          2  Technical Degree   
3                       NaN                 2          4           Medical   
4                       NaN                 3          3  Technical Degree   

   employeecount  employeenumber  environmentsatisfaction  gender  hourlyrate  \
0              1               1                        1       0         NaN   
1              1               2                        3       0         NaN   
2              1               3                        3       0         NaN   
3              1               4                        1       1         NaN   
4              1               5                        1       1         NaN   

   jobinvolvement  joblevel              jobrole  jobsatisfaction  \
0               3         5   resEArch DIREcToR                 3   
1               2         5             ManAGeR                 3   
2               3         5             ManaGER                 4   
3               3         4   ReseArCH DIrECtOr                 3   
4               4         4     sAleS EXECUtIve                 1   

  maritalstatus monthlyincome monthlyrate  numcompaniesworked over18 overtime  \
0           NaN     16280,83$   42330,17$                   7      Y       No   
1           NaN           NaN   43331,17$                   0    NaN      NaN   
2       Married           NaN   41669,33$                   1    NaN       No   
3       Married     14307,50$   37199,50$                   3      Y      NaN   
4      Divorced     12783,92$   33238,20$                   2      Y       No   

   percentsalaryhike performancerating  relationshipsatisfaction  \
0                 13               3,0                         3   
1                 14               3,0                         1   
2                 11               3,0                         4   
3                 19               3,0                         2   
4                 12               3,0                         4   

  standardhours  stockoptionlevel totalworkingyears  trainingtimeslastyear  \
0     Full Time                 0               NaN                      5   
1           NaN                 1              34,0                      5   
2           NaN                 0              22,0                      3   
3     Full Time                 2               NaN                      2   
4           NaN                 1               NaN                      5   

  worklifebalance  yearsatcompany yearsincurrentrole  yearssincelastpromotion  \
0             3,0              20                NaN                       15   
1             3,0              33                NaN                       11   
2             NaN              22                NaN                       11   
3             NaN              20                NaN                        5   
4             3,0              19                NaN                        2   

   yearswithcurrmanager sameasmonthlyincome  datebirth      salary  \
0                    15           16280,83$       1972  195370,00$   
1                     9                 NaN       1971  199990,00$   
2                    15                 NaN       1981  192320,00$   
3                     6           14307,50$       1976  171690,00$   
4                     8           12783,92$       1977         NaN   

    

In [ ]:
print(df.shape)    #cuantas filas y columnas tiene el CSV

(1678, 42)


In [ ]:
columnas = [
    "environmentsatisfaction",
    "gender",
    "hourlyrate",
    "jobinvolvement",
    "joblevel",
    "jobrole",
    "jobsatisfaction",
    "maritalstatus",
    "monthlyincome",
    "monthlyrate"
]

df_filtrado = df[columnas] #definir solo mis columnas asignadas
df_filtrado.head()

environmentsatisfaction  gender  hourlyrate  jobinvolvement  joblevel  \
0                        1       0         NaN               3         5   
1                        3       0         NaN               2         5   
2                        3       0         NaN               3         5   
3                        1       1         NaN               3         4   
4                        1       1         NaN               4         4   

               jobrole  jobsatisfaction maritalstatus monthlyincome  \
0   resEArch DIREcToR                 3           NaN     16280,83$   
1             ManAGeR                 3           NaN           NaN   
2             ManaGER                 4       Married           NaN   
3   ReseArCH DIrECtOr                 3       Married     14307,50$   
4     sAleS EXECUtIve                 1      Divorced     12783,92$   

  monthlyrate  
0   42330,17$  
1   43331,17$  
2   41669,33$  
3   37199,50$  
4   33238,20$

In [ ]:
df_filtrado.info() #tipos de datos y muestra valores no nulos
df_filtrado.dtypes  #tipos de datos. Aqui  ves sin son categoricos,numericos o texto.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1678 entries, 0 to 1677
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   environmentsatisfaction  1678 non-null   int64  
 1   gender                   1678 non-null   object 
 2   hourlyrate               411 non-null    float64
 3   jobinvolvement           1678 non-null   int64  
 4   joblevel                 1678 non-null   int64  
 5   jobrole                  1678 non-null   object 
 6   jobsatisfaction          1678 non-null   int64  
 7   maritalstatus            1003 non-null   object 
 8   monthlyincome            1189 non-null   object 
 9   monthlyrate              1678 non-null   object 
dtypes: float64(1), int64(4), object(5)
memory usage: 131.2+ KB


environmentsatisfaction      int64
gender                      object
hourlyrate                 float64
jobinvolvement               int64
joblevel                     int64
jobrole                     object
jobsatisfaction              int64
maritalstatus               object
monthlyincome               object
monthlyrate                 object
dtype: object

#descripcion completa de los datos.

| **Columna**               | **Descripción**                     | **Tipo de dato** | **Valores posibles / Ejemplos**              |
| ------------------------- | ----------------------------------- | ---------------- | -------------------------------------------- |
| `environmentsatisfaction` | Satisfacción con el entorno laboral | Numérico (1–4)   | 1 = Bajo, 2 = Medio, 3 = Alto, 4 = Muy alto  |
| `gender`                  | Género del empleado                 | Categórico       | Male / Female                                |
| `hourlyrate`              | Salario por hora                    | Numérico         | Ejemplo: 25                                  |
| `jobinvolvement`          | Nivel de implicación en el trabajo  | Numérico (1–4)   | 1 = Bajo, 2 = Medio, 3 = Alto, 4 = Muy alto  |
| `joblevel`                | Nivel del puesto                    | Numérico (1–5)   | 1 = Junior … 5 = Dirección                   |
| `jobrole`                 | Rol del empleado                    | Categórico       | Ejemplo: Sales Executive, Research Scientist |
| `jobsatisfaction`         | Satisfacción con el trabajo         | Numérico (1–4)   | 1 = Bajo, 4 = Muy alto                       |
| `maritalstatus`           | Estado civil                        | Categórico       | Single, Married, Divorced                    |
| `monthlyincome`           | Ingreso mensual                     | Numérico         | Ejemplo: 3500                                |
| `monthlyrate`             | Tarifa mensual asignada             | Numérico         | Ejemplo: 12000                               |


In [ ]:
df_filtrado.isnull().sum() #detectar los nulos de cada columna para descartar las que no sean relevantes y ahorrar trabajo.

environmentsatisfaction       0
gender                        0
hourlyrate                 1267
jobinvolvement                0
joblevel                      0
jobrole                       0
jobsatisfaction               0
maritalstatus               675
monthlyincome               489
monthlyrate                   0
dtype: int64

#hourlyrate: presenta 1267 nulos.Creo que hay que eliminarla por baja relevancia para el analisis.
#maritalstatus aunque tiene muchos nulos la dejaria porque da contexto social y aporta info sobre brecha salarial y patrones de estabilidad.(nulos serán Unknown)
#monthlyincome se conserva porque influye directamente en la satisfaccion y retencion. se puede rellenar con una media o mediana o estimar nulos a partir del salary anual divido entre 12.

In [ ]:
df_filtrado.duplicated().sum() #localiza y cuenta duplicados

np.int64(64)

In [ ]:
#lista de columnas categoricas mas sospechosas de inconsistencias y las recorrer con un bucle 

cols = ["jobrole", "gender", "maritalstatus"]

for col in cols:
    print("\n", col)
    print(df_filtrado[col].value_counts(dropna=False)) #dropna=false te dice los nulos que hay en cada categoria 



 jobrole
jobrole
mANager                       5
mAnageR                       3
ManagEr                       3
mAnaGeR                       3
ManageR                       3
MANAgER                       3
maNaGeR                       3
SALes RePREsEntAtivE          2
SALes exeCUtiVE               2
LAboRaTorY teChniCiaN         2
reseaRCH DIrectOr             2
ManUFaCtURInG dIReCToR        2
MaNAgeR                       2
mANAgER                       2
ManAgEr                       2
HEalTHcArE repREsENtaTIVE     2
resEArch scIENTisT            2
sAlES eXeCuTIVe               2
lAboratORy TeChNiCiaN         2
mAnUfactURing dIReCtOr        2
saLEs exEcUTIVe               2
MANAGER                       2
sAlES eXEcuTivE               2
maNufaCturing direcTOr        2
reSearch sCienTiSt            2
HEaltHCarE REPreSENtAtIve     2
MANuFaCtURIng DIRectoR        2
MANagER                       2
SalES ExeCuTIVe               2
MaNagER                       2
MAnagER               

In [ ]:
df_filtrado["jobrole"].value_counts() #para limpieza en valores categoricos es muy util porque dice que categoria existe 
#y tambien cuantas veces aparece cada una. Usar esta sintaxis si solo quieres consultar una columna. 

jobrole
mANager                       5
mAnageR                       3
ManagEr                       3
mAnaGeR                       3
ManageR                       3
MANAgER                       3
maNaGeR                       3
SALes RePREsEntAtivE          2
SALes exeCUtiVE               2
LAboRaTorY teChniCiaN         2
reseaRCH DIrectOr             2
ManUFaCtURInG dIReCToR        2
MaNAgeR                       2
mANAgER                       2
ManAgEr                       2
HEalTHcArE repREsENtaTIVE     2
resEArch scIENTisT            2
sAlES eXeCuTIVe               2
lAboratORy TeChNiCiaN         2
mAnUfactURing dIReCtOr        2
saLEs exEcUTIVe               2
MANAGER                       2
sAlES eXEcuTivE               2
maNufaCturing direcTOr        2
reSearch sCienTiSt            2
HEaltHCarE REPreSENtAtIve     2
MANuFaCtURIng DIRectoR        2
MANagER                       2
SalES ExeCuTIVe               2
MaNagER                       2
MAnagER                       2


la columna jobrole tiene el mismo rol escrito con muchas combinaciones de mayúsculas y minúsculas (mAnager, ManaGeR, MANAGER …) hay que normalizar con .strip() y .lower() con .str delante que permite aplicarlo a todas las filas de texto que aparezcan en la columna.

In [ ]:
df_filtrado.describe() #devuelve para cada columnas numericas(count,mean,std,min/max,25%,50%,75%)

environmentsatisfaction  hourlyrate  jobinvolvement     joblevel  \
count              1678.000000  411.000000     1678.000000  1678.000000   
mean                  4.264005   83.140768        2.740763     2.064362   
std                   6.912695   57.272101        0.710359     1.099425   
min                   1.000000   13.012897        1.000000     1.000000   
25%                   2.000000   36.254439        2.000000     1.000000   
50%                   3.000000   69.532083        3.000000     2.000000   
75%                   4.000000  116.987103        3.000000     3.000000   
max                  49.000000  255.963294        4.000000     5.000000   

       jobsatisfaction  
count      1678.000000  
mean          2.743147  
std           1.105172  
min           1.000000  
25%           2.000000  
50%           3.000000  
75%           4.000000  
max           4.000000

environmetsatisfaction tiene un rango del 1-4 asi que hay errores en valores (ejemplo 49) que hay que corregir.
hourlyrate no parece muy relevante teniendo salary y monthlyincome. Ademas está llena de nulos.(quitar) utilizando 



HourlyRate: muchos nulos y redundante, eliminar.  
MonthlyRate: redundante con DailyRate, eliminar.  
MonthlyIncome: se conserva, nulos a imputar con mediana o calculando desde Salary.  
MaritalStatus: se conserva, nulos reemplazar por Unknown y corregir errores.  
Gender: normalizado (0/1 → Male/Female), sin nulos.  
JobRole: normalizar (mayúsculas/minúsculas unificadas).  
EnvironmentSatisfaction, JobInvolvement, JobSatisfaction: escalas 1–4, corregir valores erróneos como 49 → 4.  
JobLevel: escala 1–5, correcto.  
Duplicados: se detectaron 64, eliminar.  

 #conclusiones EDA
- HourlyRate → eliminar  
- MonthlyRate → eliminar  
- MonthlyIncome → imputar  
- MaritalStatus → conservar, Unknown  
- Gender → normalizar  
- JobRole → normalizar  
- EnvSat, JobInv, JobSat → corregir  
- JobLevel → ok  
- Duplicados → eliminar (64)


### Guía rápida EDA (Exploratory Data Analysis)

- **Importar librerías y dataset**  
  - `import pandas as pd`  
  - `df = pd.read_csv("archivo.csv")`  

- **Inventario de datos**  
  - `df.columns` → nombres de columnas  
  - `df.shape` → nº de filas y columnas  
  - `df.head()` → primeras filas  

- **Tipos de datos y nulos**  
  - `df.info()` → tipo de dato + nº de valores no nulos  
  - `df.isnull().sum()` → nº de nulos por columna  

- **Duplicados**  
  - `df.duplicated().sum()` → cuántos duplicados  
  - `df = df.drop_duplicates()`  

- **Valores únicos en categóricas**  
  - `df["col"].value_counts(dropna=False)`  
  - Bucle para recorrer varias columnas categóricas sospechosas  

- **Normalización de texto**  
  - `.str.strip().str.lower()` → quitar espacios y unificar mayúsculas/minúsculas  

- **Estadísticos descriptivos (numéricas)**  
  - `df.describe()` → count, mean, std, min, max, percentiles  
  - Revisar rangos válidos (ej. escalas 1–4, 1–5, etc.)  

- **Decisiones de limpieza**  
  - Eliminar columnas irrelevantes/redundantes  
  - Imputar nulos (media/mediana/"Unknown")  
  - Normalizar categorías (ej. Male/Female)  
  - Corregir valores erróneos  

- **Diccionario de datos limpio**  
  - Documentar columna, tipo de dato, descripción y decisión tomada  
